File: generated_code/GeminiPro/Java/076_SplayTree.java
import java.util.NoSuchElementException;

/**
 * A Splay Tree implementation that stores a set of integers.
 *
 * A splay tree is a self-adjusting binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, look-up and removal
 * in O(log n) amortized time.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a left rotation on the given node x.
     *
     *      x               y
     *     / \             / \
     *    A   y    -->    x   C
     *       / \         / \
     *      B   C       A   B
     *
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node x.
     *
     *        x           y
     *       / \         / \
     *      y   C  -->  A   x
     *     / \             / \
     *    A   B           B   C
     *
     * @param x The node to rotate.
     */
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y != null) {
            x.left = y.right;
            if (y.right != null) {
                y.right.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.right = x;
            x.parent = y;
        }
    }

    /**
     * Moves the given node x to the root of the tree using a series of rotations.
     *
     * @param x The node to move to the root.
     */
    private void splay(Node x) {
        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;

            if (grandParent == null) { // Zig case
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else {
                if (x == parent.left && parent == grandParent.left) { // Zig-Zig case (left-left)
                    rightRotate(grandParent);
                    rightRotate(parent);
                } else if (x == parent.right && parent == grandParent.right) { // Zig-Zig case (right-right)
                    leftRotate(grandParent);
                    leftRotate(parent);
                } else if (x == parent.right && parent == grandParent.left) { // Zig-Zag case (left-right)
                    leftRotate(parent);
                    rightRotate(grandParent);
                } else { // Zig-Zag case (right-left)
                    rightRotate(parent);
                    leftRotate(grandParent);
                }
            }
        }
        this.root = x;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the node with that key is splayed to the root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node current = root;
        Node parent = null;
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay the existing node and return
                splay(current);
                return;
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }

        // Splay the new node to the root
        splay(newNode);
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, the corresponding node is splayed to the root.
     * If the key is not found, the last accessed node is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        Node current = root;
        Node lastAccessed = null;

        while (current != null) {
            lastAccessed = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key found, splay the node and return the key
                splay(current);
                return current.key;
            }
        }

        // Key not found, splay the last accessed node (if any)
        if (lastAccessed != null) {
            splay(lastAccessed);
        }
        return null;
    }

    /**
     * Deletes a key from the splay tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        // Search for the key. This will splay the node to the root if it exists,
        // or the last-accessed node if it doesn't.
        search(key);

        // If the key is not in the tree, or the tree is empty, do nothing.
        // After search, if the key was found, it must be at the root.
        if (root == null || root.key != key) {
            return;
        }

        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // No left child, the right subtree becomes the new tree
            root = rightSubtree;
            if (root != null) {
                root.parent = null;
            }
        } else {
            // Disconnect the left subtree from the old root
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }

            // Splay this max node. It becomes the new root of the combined tree.
            // We temporarily set the root to the left subtree to perform the splay within it.
            this.root = leftSubtree;
            splay(maxNode);

            // The new root is now maxNode. Attach the original right subtree.
            // After splaying, maxNode (now the root) has no right child.
            root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = root;
            }
        }
    }
}

File: generated_code/GeminiPro/Python/076_SplayTree.py
class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree.

    This class provides a dictionary-like set for storing unique integer keys.
    It supports insertion, deletion, and search operations. The key feature
    of a splay tree is that recently accessed elements are moved to the root
    of the tree, which can improve performance for certain access patterns.

    Methods:
        insert(key): Inserts a key into the tree.
        delete(key): Deletes a key from the tree.
        search(key): Searches for a key and splays the accessed node.
    """

    class Node:
        """A node in the splay tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    # --- Private Helper Methods for Tree Rotations and Splaying ---

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left is not None:
            y.left.parent = x

        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right is not None:
            y.right.parent = x

        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splay operation on a node, moving it to the root.
        """
        while node.parent is not None:
            parent = node.parent
            grandparent = parent.parent
            if grandparent is None:  # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    # --- Public API Methods ---

    def search(self, key):
        """
        Searches for a key in the tree.

        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last accessed node (the parent of the
        would-be key) is splayed to the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        node = self.root
        last_visited = None
        while node is not None:
            last_visited = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        # Key not found, splay the last visited node if tree is not empty
        if last_visited is not None:
            self._splay(last_visited)
        return False

    def insert(self, key):
        """
        Inserts a key into the splay tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        # Find position for new node (standard BST insert)
        parent = None
        current = self.root
        while current is not None:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay it and return
                self._splay(current)
                return

        # Insert the new node
        new_node = self.Node(key, parent)
        if parent is None:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the newly inserted node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the splay tree.

        The node with the key (or its parent if not found) is first splayed
        to the root. If the key is found at the root, it is removed, and the
        remaining subtrees are joined.

        Args:
            key (int): The integer key to delete.
        """
        # search() will splay the node or its parent to the root
        if not self.search(key):
            # Key was not in the tree. search() splayed the closest node, so we're done.
            return

        # If search returned True, the node to delete is now the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if left_subtree is None:
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif right_subtree is None:
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Both subtrees exist. Join them.
            # Detach the left subtree to operate on it.
            left_subtree.parent = None

            # Find the maximum node in the left subtree.
            max_in_left = left_subtree
            while max_in_left.right is not None:
                max_in_left = max_in_left.right
            
            # Splay this max node to the root of its subtree.
            # We temporarily set self.root to the left subtree's root to
            # perform the splay operation within that subtree.
            self.root = left_subtree
            self._splay(max_in_left)
            # After splaying, self.root is max_in_left, which becomes the new
            # root of the combined tree.
            
            # Attach the original right subtree to the new root.
            self.root.right = right_subtree
            right_subtree.parent = self.root